### PR TITLE
Integration of rack-timeout for limiting request time

### PR DESCRIPTION
**Pull request by ajaystha42**
Addresses: https://github.com/restarone/violet_rails/issues/1397

Demo: 

https://user-images.githubusercontent.com/55496123/217051621-69b8e185-31e3-4288-b533-383046e7fd2e.mov

### CODE REVIEW

**Comment by ajaystha42** [comment]
@donrestarone  There were some issues when bundle installing the new gem in the application. Thats why I only kept the changes related to rack-timeout gem.

File: .env.test [hunk 1]
```diff
@@ -9,4 +9,4 @@ APP_HOST=lvh.me:5250
 RECAPTCHA_SITE_KEY=6Lc6BAAAAAAAAChqRbQZcn_yyyyyyyyyyyyyyyyy
 RECAPTCHA_SECRET_KEY=6Lc6BAAAAAAAAKN3DRm6VA_xxxxxxxxxxxxxxxxx
 SECRET_KEY_BASE='38c72586473e364229897f24f1892f1dc5565776878aa4d8c6bf051258622bd2e923b926ab59b40f912b661216f764d993e8d6b8bbfbc33026e5c954b6c51f9b'
-
+RACK_TIMEOUT_SERVICE_TIMEOUT=8
```
Comment (donrestarone) [inline .env.test:12]: this is unused. When you use it please rename it to `RACK_TIMEOUT`

Comment (donrestarone) [reply .env.test:12]: nvm, this is used via automagic`RACK_TIMEOUT_SERVICE_TIMEOUT`

**Comment by donrestarone** [comment]
@ajaystha42 please add a controller test that sets up a custom action with a `sleep` longer than the default timeout (in seconds) and assert that the request is interrupted by the rack timeout

**Comment by donrestarone** [comment]
also, you need a demo video and a reference to the GitHub issue in the description

**Comment by ajaystha42** [comment]
> also, you need a demo video and a reference to the GitHub issue in the description

What should the demo video cover for this task?  @donrestarone

**Comment by donrestarone** [comment]
@ajaystha42 the demo should show that requests taking longer than the default timeout are timed out in the default timeout duration

File: test/controllers/admin/comfy/api_clients_controller_test.rb [hunk 2]
```diff
@@ -718,5 +719,124 @@ class Comfy::Admin::ApiClientsControllerTest < ActionDispatch::IntegrationTest
...
+  #   end
+  # end
+
+
+  # test 'test_rack_timeout_should_throw_timed_out_exception' do
+  #   Rack::Timeout.stubs(:timeout).returns(0.0001)
+
+  #   MyModule::MySinatra.with_fake_route(:get, '/dosomething', ->{ sleep 0.0002 }) do
+  #     get '/dosomething'
+  #   end
```
Comment (donrestarone) [inline test/controllers/admin/comfy/api_clients_controller_test.rb:748]: @ajaystha42 any updates on this PR? If its ready for UAT you should remove this commented out code.

**Comment by Pralish** [comment]
@ajaystha42 , Here's the rack timeout integration test case. You can add more edge cases if you.

**Comment by donrestarone** [comment]
⚠️ temporary deployment to markedrestaurant.com for testing

**Comment by donrestarone** [comment]
⚠️ temporary deployment to markedrestaurant.com for testing

https://github.com/restarone/violet_rails/settings/environments/442907780/edit


UPDATE:

Rack timeout is sending error email notification with the relevant data, but when I tested the server crashed right after the exception notification was sent. So I believe we are not recovering from the timeout correctly.



``` txt

A Rack::Timeout::RequestTimeoutException occurred in dashboard#dashboard:

  Request ran for longer than 15000ms 
  app/helpers/dashboard_helper.rb:27:in `block in page_visit_chart_data'


-------------------------------
Request:
-------------------------------

  * URL        : https://markedrestaurant.com/v2/dashboard
  * HTTP Method: GET
  * IP address : 74.15.198.4
  * Parameters : {"controller"=>"comfy/admin/v2/dashboard", "action"=>"dashboard"}
  * Timestamp  : 2023-02-17 03:24:54 UTC
  * Server : ip-172-31-18-133
    * Rails root : /var/www/violet/releases/20230217031903
  * Process: 1662862

-------------------------------
Session:
-------------------------------

  * session id: [FILTERED]
  * data: {"session_id"=>"f11faf60d23b8a34b46d67c35a2267d8",
   "user_return_to"=>"/v2/dashboard",
   "warden.user.user.key"=>[[1], "$2a$12$"],
   "warden.user.user.session"=>{"last_request_at"=>1676604279},
   "site_id"=>1,
   "_csrf_token"=>"efaeferf="}

-------------------------------
Environment:
-------------------------------

  * GATEWAY_INTERFACE                                       : CGI/1.2
    * HTTP_ACCEPT                                             : text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
    * HTTP_ACCEPT_ENCODING                                    : gzip, deflate, br
    * HTTP_ACCEPT_LANGUAGE                                    : en-CA,en-US;q=0.9,en;q=0.8
    * HTTP_CONNECTION                                         : close
    * HTTP_COOKIE                                             : _r_solutions_session=LXyq07%2Bt4cqWpluJ5NDZPgM600oEgwBR7XiDA5VmkTGTBKnRhDDdLt9UBCbt5Irs955AmHmcJ9zqFSK4uvsitMRfqSoRkSG9Xh3Cx%2F5CiWmsaSYK8I1d1O3i1RjsA5uJzSZUVuX7sl7mIsjzEtuMpa3HfrlLcafa7ooOTsBb%2BtCoLsFrQNooQsrOOPWH7x0Ym8oxsK7og47kyckJejMW8AYAcvG7fThAcjEp2Zc3ohHFY7MM13ooerZabcXKr1uzu9i2Npu565SMsbYypEzImv%2FMCfMZwi9ox5e9axdFZ0mWuQ4Y1aDcTx1MfiT4eCl%2F1ymc2d23fa%2FF2tgcLChGC9DRg5raYXYSyWLJ80PvMQS6qIBeazJO%2BxR4Pdgk6alsGsW48XXDXqHF%2BxKWgthKIXM0hs4YMhhBdW8LCAMrdmRnen3iH59rc4SqVgzvV%2BUTV1N5iD1sBEfge3DA0dgZvweadSmnx2ppFFe04b%2B%2Ffx16XspPPgyztvvThDoSLsM4pqyp7TXP74jX%2FmVsS2A08R6bpTOpeab%2FhTnE3E5HHLZhQBwpcJ96NQrDSrhTakSXDM0f%2BaaTaIvNl0fPDP7mMg%3D%3D--%2FDDrOio08QsYRjVp--i7IE2JD9rBHtpWk4bbomXQ%3D%3D; ahoy_visit=6727a696-20ad-47a1-92f0-35b5676f460b; ahoy_visitor=fd508a7c-0900-462a-88cb-7961355993d2; cookies_accepted=true
    * HTTP_HOST                                               : [markedrestaurant.com](http://markedrestaurant.com/)
    * HTTP_REFERER                                            : https://markedrestaurant.com/admin/sites/1/pages
    * HTTP_USER_AGENT                                         : Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Safari/605.1.15
    * HTTP_VERSION                                            : HTTP/1.0
    * HTTP_X_FORWARDED_FOR                                    : 74.15.198.4
    * HTTP_X_FORWARDED_PROTO                                  : https
    * HTTP_X_REAL_IP                                          : 74.15.198.4
    * ORIGINAL_FULLPATH                                       : /v2/dashboard
    * ORIGINAL_SCRIPT_NAME                                    :
    * PATH_INFO                                               : /v2/dashboard
    * QUERY_STRING                                            :
    * REMOTE_ADDR                                             : 127.0.0.1
    * REQUEST_METHOD                                          : GET
    * REQUEST_PATH                                            : /v2/dashboard
    * REQUEST_URI                                             : /v2/dashboard
    * ROUTES_47008777893880_SCRIPT_NAME                       :
    * SCRIPT_NAME                                             :
    * SERVER_NAME                                             : [markedrestaurant.com](http://markedrestaurant.com/)
    * SERVER_PORT                                             : 443
    * SERVER_PROTOCOL                                         : HTTP/1.1
    * SERVER_SOFTWARE                                         : puma 5.6.4 Birdie's Version
    * action_controller.instance                              : #<Comfy::Admin::V2::DashboardController:0x00007f14509de480>
    * action_dispatch.authenticated_encrypted_cookie_salt     : [FILTERED]
    * action_dispatch.backtrace_cleaner                       : #<Rails::BacktraceCleaner:0x000055822bdcc398>
    * action_dispatch.content_security_policy                 :
    * action_dispatch.content_security_policy_nonce_directives:
    * action_dispatch.content_security_policy_nonce_generator :
    * action_dispatch.content_security_policy_report_only     : false
    * action_dispatch.cookies                                 : #<ActionDispatch::Cookies::CookieJar:0x00007f14507c2b38>
    * action_dispatch.cookies_digest                          :
    * action_dispatch.cookies_rotations                       : #<ActiveSupport::Messages::RotationConfiguration:0x00005582273af508>
    * action_dispatch.cookies_same_site_protection            : #<Proc:0x00005582283af840@/var/www/violet/shared/bundle/ruby/2.6.0/gems/railties-6.1.5/lib/rails/application.rb:636>
    * action_dispatch.cookies_serializer                      : json
    * action_dispatch.encrypted_cookie_cipher                 : [FILTERED]
    * action_dispatch.encrypted_cookie_salt                   : [FILTERED]
    * action_dispatch.encrypted_signed_cookie_salt            : [FILTERED]
    * action_dispatch.http_auth_salt                          : [FILTERED]
    * action_dispatch.key_generator                           : #<ActiveSupport::CachingKeyGenerator:0x00005582299f1a18>
    * action_dispatch.logger                                  : #<ActiveSupport::Logger:0x000055822ad8a6b0>
    * action_dispatch.parameter_filter                        : [:passw, :secret, :token, :_key, :crypt, :salt, :certificate, :otp, :ssn]
    * action_dispatch.permissions_policy                      :
    * action_dispatch.redirect_filter                         : []
    * action_dispatch.remote_ip                               : 74.15.198.4
    * action_dispatch.request.content_type                    :
    * action_dispatch.request.formats                         : [#<Mime::Type:0x00005582274e1908 @synonyms=["application/xhtml+xml"], @symbol=:html, @string="text/html", @hash=-4557016495958642182>]
    * action_dispatch.request.parameters                      : {"controller"=>"comfy/admin/v2/dashboard", "action"=>"dashboard"}
    * action_dispatch.request.path_parameters                 : {:controller=>"comfy/admin/v2/dashboard", :action=>"dashboard"}
    * action_dispatch.request.query_parameters                : {}
    * action_dispatch.request.request_parameters              : {}
    * action_dispatch.request.unsigned_session_cookie         : {"session_id"=>"f11faf60d23b8a34b46d67c35a2267d8", "user_return_to"=>"/admin/sites/1/pages", "warden.user.user.key"=>[[1], "$2a$12$KTdbOzd3ZXy7QlSSwsY90e"], "warden.user.user.session"=>{"last_request_at"=>1676604279}, "site_id"=>1, "_csrf_token"=>"[FILTERED]"}
    * action_dispatch.request_id                              : ffd3a478-7bad-40b5-a5b6-4cf7c8d7a4fc
    * action_dispatch.routes                                  : #<ActionDispatch::Routing::RouteSet:0x000055822b03eff0>
    * action_dispatch.secret_key_base                         : [FILTERED]
    * action_dispatch.show_detailed_exceptions                : false
    * action_dispatch.show_exceptions                         : true
    * action_dispatch.signed_cookie_digest                    :
    * action_dispatch.signed_cookie_salt                      : [FILTERED]
    * action_dispatch.use_authenticated_cookie_encryption     : [FILTERED]
    * action_dispatch.use_cookies_with_metadata               : true
    * exception_notifier.exception_data                       : {:current_user=>#<User id: 1, global_admin: true, created_at: "2021-06-14 00:37:04.410710000 +0000", updated_at: "2023-02-17 03:24:36.714734000 +0000", email: "[contact@restarone.com](mailto:contact@restarone.com)", can_manage_web: true, can_manage_email: true, can_manage_users: true, can_manage_blog: true, name: "Shashike Jayatun...
    * puma.config                                             : #<Puma::Configuration:0x0000558225a626b8>
    * puma.request_body_wait                                  : 0
    * puma.socket                                             : #<TCPSocket:0x00007f14507497d8>
    * [rack-timeout.info](http://rack-timeout.info/)                                       : #<struct Rack::Timeout::RequestDetails id="dd4f97fa-c5c2-4e0a-8a5d-eec521c024af", wait=nil, service=15.000879144994542, timeout=15, state=:timed_out, term=false>
    * rack.after_reply                                        : []
    * rack.cors                                               : #<Rack::Cors::Result:0x00007f1450749030>
    * rack.errors                                             : #<IO:0x00005582250b80b8>
    * rack.hijack                                             : #<Puma::Client:0x00007f14507497b0>
    * rack.hijack?                                            : true
    * rack.input                                              : #<Puma::NullIO:0x0000558225cda8a0>
    * rack.multiprocess                                       : false
    * rack.multithread                                        : true
    * rack.request.cookie_hash                                : {"_r_solutions_session"=>"LXyq07+t4cqWpluJ5NDZPgM600oEgwBR7XiDA5VmkTGTBKnRhDDdLt9UBCbt5Irs955AmHmcJ9zqFSK4uvsitMRfqSoRkSG9Xh3Cx/5CiWmsaSYK8I1d1O3i1RjsA5uJzSZUVuX7sl7mIsjzEtuMpa3HfrlLcafa7ooOTsBb+tCoLsFrQNooQsrOOPWH7x0Ym8oxsK7og47kyckJejMW8AYAcvG7fThAcjEp2Zc3ohHFY7MM13ooerZabcXKr1uzu9i2Npu565SMsbYypE...
    * rack.request.cookie_string                              : _r_solutions_session=LXyq07%2Bt4cqWpluJ5NDZPgM600oEgwBR7XiDA5VmkTGTBKnRhDDdLt9UBCbt5Irs955AmHmcJ9zqFSK4uvsitMRfqSoRkSG9Xh3Cx%2F5CiWmsaSYK8I1d1O3i1RjsA5uJzSZUVuX7sl7mIsjzEtuMpa3HfrlLcafa7ooOTsBb%2BtCoLsFrQNooQsrOOPWH7x0Ym8oxsK7og47kyckJejMW8AYAcvG7fThAcjEp2Zc3ohHFY7MM13ooerZabcXKr1uzu9i2Npu565SMsbYypEzImv%2FMCfMZwi9ox5e9axdFZ0mWuQ4Y1aDcTx1MfiT4eCl%2F1ymc2d23fa%2FF2tgcLChGC9DRg5raYXYSyWLJ80PvMQS6qIBeazJO%2BxR4Pdgk6alsGsW48XXDXqHF%2BxKWgthKIXM0hs4YMhhBdW8LCAMrdmRnen3iH59rc4SqVgzvV%2BUTV1N5iD1sBEfge3DA0dgZvweadSmnx2ppFFe04b%2B%2Ffx16XspPPgyztvvThDoSLsM4pqyp7TXP74jX%2FmVsS2A08R6bpTOpeab%2FhTnE3E5HHLZhQBwpcJ96NQrDSrhTakSXDM0f%2BaaTaIvNl0fPDP7mMg%3D%3D--%2FDDrOio08QsYRjVp--i7IE2JD9rBHtpWk4bbomXQ%3D%3D; ahoy_visit=6727a696-20ad-47a1-92f0-35b5676f460b; ahoy_visitor=fd508a7c-0900-462a-88cb-7961355993d2; cookies_accepted=true
    * rack.request.query_hash                                 : {}
    * rack.request.query_string                               :
    * rack.run_once                                           : false
    * rack.session                                            : #<ActionDispatch::Request::Session:0x00007f14508c2010>
    * rack.session.options                                    : #<ActionDispatch::Request::Session::Options:0x00007f14508c1fc0>
    * rack.tempfiles                                          : []
    * rack.url_scheme                                         : https
    * rack.version                                            : [1, 6]
    * warden                                                  : Warden::Proxy:69862613716840 @config={:default_scope=>:user, :scope_defaults=>{}, :default_strategies=>{:user=>[:two_factor_authenticatable, :two_factor_authenticatable, :rememberable]}, :intercept_401=>false, :failure_app=>RSolutionsFailureApp}

-------------------------------
Backtrace:
-------------------------------

  app/helpers/dashboard_helper.rb:27:in `block in page_visit_chart_data'
  app/helpers/dashboard_helper.rb:27:in `group_by'
  app/helpers/dashboard_helper.rb:27:in `page_visit_chart_data'
  app/views/comfy/admin/v2/dashboard/dashboard.html.haml:41

-------------------------------
Data:
-------------------------------

  * data: {:current_user=>
    #<User id: 1, global_admin: true, created_at: "2021-06-14 00:37:04.410710000 +0000", updated_at: "2023-02-17 03:24:36.714734000 +0000", email: "[contact@restarone.com](mailto:contact@restarone.com)", can_manage_web: true, can_manage_email: true, can_manage_users: true, can_manage_blog: true, name: "Shashike Jayatunge", moderator: true, can_view_restricted_pages: true, deliver_analytics_report: false, can_manage_subdomain_settings: true, session_timeoutable_in: "1-week", can_access_admin: true, deliver_error_notifications: true, can_manage_analytics: true, can_manage_files: true, encrypted_otp_secret: nil, encrypted_otp_secret_iv: nil, encrypted_otp_secret_salt: nil, consumed_timestep: nil, otp_required_for_login: false, api_accessibility: {"api_keys"=>{"full_access"=>"true"}, "api_namespaces"=>{"all_namespaces"=>{"full_access"=>"true"}}}, can_access_forum: true, otp_secret: nil>,
   :current_visit=>
    #<Ahoy::Visit:0x00007f1450c476e0
     id: 678521,
     visit_token: "6727a696-20ad-47a1-92f0-35b5676f460b",
     visitor_token: "fd508a7c-0900-462a-88cb-7961355993d2",
     user_id: 1,
     ip: "74.15.198.4",
     user_agent:
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Safari/605.1.15",
     referrer: nil,
     referring_domain: nil,
     landing_page: "https://markedrestaurant.com/",
     browser: "Safari",
     os: "Mac",
     device_type: "Desktop",
     country: nil,
     region: nil,
     city: nil,
     latitude: nil,
     longitude: nil,
     utm_source: nil,
     utm_medium: nil,
     utm_term: nil,
     utm_content: nil,
     utm_campaign: nil,
     app_version: nil,
     os_version: nil,
     platform: nil,
     started_at: Fri, 17 Feb 2023 03:24:25.994988000 UTC +00:00>}
```

File: test/integration/rack_timeout_test.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+class RackTimeoutTest < ActionDispatch::IntegrationTest
+  setup do
+    setup_fake_routes
+  end
+
+  teardown do
+    Rails.application.reload_routes!
+  end
+
```
Comment (donrestarone) [inline test/integration/rack_timeout_test.rb:11]: furthering my comment here: https://github.com/restarone/violet_rails/pull/1402#issuecomment-1434052107
can we add a test to validate that after the request is aborted the server is still heathy? To do verify health, we will need to do the following in the test,

1. Make a request that gets aborted
2. assert that the request was aborted
3. make a request to the server and ensure that it can handle web requests and background jobs


CC @Pralish @ajaystha42

**Comment by donrestarone** [comment]
⚠️  temporary deployment to markezrestaurant.com: https://github.com/restarone/violet_rails/settings/environments/442907780/edit
